test/test_solvers.jl

```diff
@@ -296,22 +297,169 @@ function can_solve_batched_tridiagonal_system_with_3D_functions(arch, Nx, Ny, Nz
     return Array(ϕ) ≈ ϕ_correct
 end
 
+function vertically_stretched_grid_div_free_velocity(arch, Nx, Ny, zF)
```

Comment:
It looks like this test solves the Poisson equation with an eigenfunction expansion. Is this correct? I don't think it is a literal test of the enforcement of a divergence free velocity field via the fractional step algorithm. If so, I think the name of this test should be renamed. Also, much of the complexity associated with computing the divergence of some velocity field, etc can be avoided. I think we can just randomly construct `R`, compute the solution `ϕ` to the Poisson equation `∇²ϕ = R`, and then compare `R` to `∇²ϕ`. Right?

Reply:
This test could also use a docstring I think because it is quite complicated --- may be hard to follow for future developers. I know many of our tests don't have docstrings, but maybe we should change that?

Reply:
> It looks like this test solves the Poisson equation with an eigenfunction expansion. Is this correct? I don't think it is a literal test of the enforcement of a divergence free velocity field via the fractional step algorithm. If so, I think the name of this test should be renamed.

That's a good point. I renamed the test to `vertically_stretched_poisson_solver_correct_answer`.

> Also, much of the complexity associated with computing the divergence of some velocity field, etc can be avoided. I think we can just randomly construct `R`, compute the solution `ϕ` to the Poisson equation `∇²ϕ = R`, and then compare `R` to `∇²ϕ`. Right?

That's exactly what the test used to do but I think the right-hand side needs to actually satisfy the boundary condition $G_u \dot \hat{n} = 0$ according to Marshall, Hill, Perelman, & Adcroft (section 3.2, 1997).

If I just give it a random right hand side with zero mean the test fails. I believe that the discrete cosine transform was enforcing this boundary condition for us but now we actually have to enforce us ourselves, but not sure how to show this.

Reply:
> If I just give it a random right hand side with zero mean the test fails.

Erm... `R` doesn't have boundary values... ? I don't think its true that `R` needs to satisfy some criterion for the test to be valid. You have have to fill the halos for `ϕ` before computing `∇²ϕ`... ? Or, perhaps you can circumvent the issue of calculating the Laplacian across the boundary by only comparing `∇²ϕ` to `R` on interior points between 2:Nx-1... ?

Reply:
> That's a good point. I renamed the test to vertically_stretched_poisson_solver_correct_answer.

Ok, but isn't this actually a test of the *tridiagonal* solver on a uniform grid?

Reply:
I am filling in the halos properly. I don't think we should shortcut the test and not check the boundaries, getting the correct solution at the boundary is an important part of the test.

Using a completely random right-hand-side with zero mean, you end up with the correct solution everywhere except at the last vertical grid point where there is a systematic error that is the same at all the columns, so I feel like there may be a simple bug somewhere that is somehow not present when $G_u \dot \hat{n} = 0$.